File: book/fridgets/c#/Fridgets/FrView.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using Sodium.Frp;
using Sodium.Functional;

namespace Fridgets
{
    public class FrView : Canvas, IDisposable
    {
        private readonly IListener l;
        private readonly Cell<DrawableDelegate> drawable;

        public FrView(Window window, Fridget fr, IListener l)
        {
            StreamSink<MouseEvent> sMouse = Stream.CreateSink<MouseEvent>();
            StreamSink<KeyEvent> sKey = Stream.CreateSink<KeyEvent>();
            this.MouseDown += (sender, args) => sMouse.Send(new MouseEvent(args, () => args.GetPosition(this)));
            this.MouseUp += (sender, args) => sMouse.Send(new MouseEvent(args, () => args.GetPosition(this)));
            this.MouseMove += (sender, args) => sMouse.Send(new MouseEvent(args, () => args.GetPosition(this)));
            CellSink<Maybe<Size>> size = Cell.CreateSink(Maybe<Size>.None);
            this.SizeChanged += (sender, args) => size.Send(Maybe.Some(args.NewSize));
            window.KeyDown += (sender, args) =>
            {
                Key key = args.Key == Key.System ? args.SystemKey : args.Key;
                if (key == Key.Back)
                {
                    sKey.Send(new BackspaceKeyEvent());
                }
            };
            window.TextInput += (sender, args) => sKey.Send(new StringKeyEvent(args.Text));
            CellLoop<long> focus = new CellLoop<long>();
            Fridget.Output fo = fr.Reify(size, sMouse, sKey, focus, new Supply());
            focus.Loop(fo.SChangeFocus.Hold(-1));
            this.drawable = fo.Drawable;
            this.l = Listener.CreateComposite(new[] { l, this.drawable.Updates().Listen(d => this.InvalidateVisual()) });
        }

        protected override void OnRender(DrawingContext dc)
        {
            base.OnRender(dc);

            this.drawable.Sample()(dc);
        }

        public void Dispose()
        {
            this.l.Unlisten();
        }
    }
}

File: book/fridgets/f#/Fridgets/FrView.fs
namespace Fridgets

open System
open System.Windows
open System.Windows.Controls
open System.Windows.Input
open Sodium.Frp

type FrView(window : Window, fr : IFridget, l : IListener) as this =
    inherit Canvas()

    let init () =
        let flip f x y = f y x
        let sMouse = sinkS ()
        let sKey = sinkS ()
        let getMouseEvent args = { args = args; getPosition = (fun () -> args.GetPosition(this)) }
        this.MouseDown.Subscribe (getMouseEvent >> (sMouse |> flip sendS)) |> ignore
        this.MouseUp.Subscribe (getMouseEvent >> (sMouse |> flip sendS)) |> ignore
        this.MouseMove.Subscribe (getMouseEvent >> (sMouse |> flip sendS)) |> ignore
        let size = sinkC None
        let getNewSize (args : SizeChangedEventArgs) = Some args.NewSize
        this.SizeChanged.Subscribe (getNewSize >> (size |> flip sendC)) |> ignore
        let onKeyDown (args : KeyEventArgs) =
            let key = if args.Key = Key.System then args.SystemKey else args.Key
            if key = Key.Back then sKey |> sendS BackspaceKeyEvent
        window.KeyDown.Subscribe onKeyDown |> ignore
        let getStringKeyEvent (args : TextCompositionEventArgs) = StringKeyEvent args.Text
        window.TextInput.Subscribe (getStringKeyEvent >> (sKey |> flip sendS)) |> ignore
        let struct (_, fo) = loopC (fun focus ->
            let fo = Fridget.reify fr size sMouse sKey focus (Supply.create ())
            struct (fo.sChangeFocus |> holdS -1L, fo))
        let drawable = fo.drawable
        let l = Listener.fromSeq [ l; upcast (drawable |> updatesC |> listenS (fun _ -> this.InvalidateVisual())) ]
        (l, drawable)
    let (l, drawable) = init ()

    override __.OnRender(dc) =
        base.OnRender(dc)
        drawable |> sampleC |> DrawableDelegate.invoke dc

    interface IDisposable with
        member __.Dispose() = l.Unlisten ()